(docs) - Add Comparison page (#1043)
FormidableLabs/urql: docs/comparison.md

| File Uploads                               | ✅ `@urql/exchange-multipart-fetch` | 🟡 `apollo-upload-client`                     | 🛑                             |
| Retrying Failed Queries                    | ✅ `@urql/exchange-retry`           | ✅ `apollo-link-retry`                        | ✅ `DefaultNetworkLayer`       |
| Easy Authentication Flows                  | ✅ `@urql/exchange-auth`            | 🛑 (no docs for refresh-based authentication) | 🟡 `react-relay-network-layer` |
| Automatic Refetch after Mutation           | ✅ (with document cache)            | 🛑                                            | ✅                             |

Typically these are all additional addon features that you may expect from a GraphQL client, no
matter which framework you use it with. It's worth mentioning that all three clients support some
kind of extensibility API which allows you to change when and how queries are sent to an API. These
are easy to use primitives particularly in Apollo, with links, and in `urql` with exchanges. The
major difference in `urql` is that all caching logic is abstracted in exchanges too, which makes

> them slightly more powerful (and hence makes `urql` slightly more customizable.)

Reviewer: I'd refrain from using powerful and just stick to "more customizable"

A lot of the added exchanges for persisted queries, file uploads, retrying, and other features are

> implemented in the core library, while there are some cases where first-party support isn't provided

Reviewer: ```suggestion
implemented by the urql-team, while there are some cases where first-party support isn't provided
```

It's not really part of the core-library

implemented by the urql-team, while there are some cases where first-party support isn't provided
in Relay or Apollo. This doesn't mean that these features can't be used with these clients, but that
you'd have to lean on community libraries or maintaining/implementing them yourself.

### Framework Bindings

> | ---                           | urql                             | Apollo             | Relay             |

Reviewer 2: the order of the columns changed - maybe keep them consistent?

Reviewer 2: you're right - wrapping in the github UI

extensive, but we attempt to lean on good and well funded maintenance at Formidable, active
investment, and good documentation. So you may have to evaluate whether you can get any questions
answered quickly enough on [GitHub Discussions](https://github.com/FormidableLabs/urql/discussions)
for instance.

### Caching and State

| ---                                                     | urql                                                                  | Apollo             | Relay                                          |
| ------------------------------------------------------- | --------------------------------------------------------------------- | ------------------ | ---------------------------------------------- |
| Caching Strategy                                        | Document Caching, Normalized Caching with `@urql/exchange-graphcache` | Normalized Caching | Normalized Caching (schema restrictions apply) |
| Added Bundle Size                                       | +5.9kB (with Graphcache)                                              | +0 (default)       | +0 (default)                                   |

> | Automatic Garbage Collection                            | ✅                                                                    | ✅                 | ✅                                             |

Reviewer: Is it automatic in apollo, thought you had to call .gc

Author: That's true, I just checked and it _isn't_ always automatic, I suppose to not break legacy apps, so I'll update this

| Pagination Support                                      | 🔶                                                                    | 🔶                 | ✅                                             |
| Optimistic Updates                                      | ✅                                                                    | ✅                 | ✅                                             |
| Local Updates                                           | ✅                                                                    | ✅                 | ✅                                             |
| Out-of-band Cache Updates                               | 🛑 (stays true to server data)                                        | ✅                 | 🛑                                             |
| Local Resolvers and Redirects                           | ✅                                                                    | ✅                 | 🛑                                             |
| Complex Resolvers (nested non-normalized return values) | ✅                                                                    | 🛑                 | 🛑                                             |

> | Commutativity Guarantees                                | ✅                                                                    | 🛑                 | 🛑                                             |

Reviewer: Could link to under-the-hood to explain this

Author: It's explained in the explanation after the table

| Partial Results                                         | ✅                                                                    | ✅                 | 🛑                                             |

> | Partial Results                                         | ✅                                                                    | ✅                 | 🛑                                             |

Reviewer: Partial results is duplicated her
